### PR TITLE
Make NULLs comparable to everything.

### PR DESCRIPTION
**Pull request by zlobober**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

In this PR we make NULL fields comparable to any other fields. Default comparison order is similar to NULLS_LAST, but it may be overridden during linkage by rewriting extern const bool DB::NULLS_LAST weak symbol.

This is important to third-party applications using ClickHouse as a library which rely on DB::KeyCondition in order to prune input possibly containing NULLs.

Changelog category (leave one):
- Non-significant (changelog entry is not required)

Related: #5319

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
> it may be overridden during linkage by rewriting extern const bool DB::NULLS_LAST weak symbol

If it is unused right now, better not introduce this tweak at all.

File: dbms/src/Common/FieldVisitors.h [hunk 1]
```diff
@@ -313,7 +328,7 @@ class FieldVisitorAccurateLess : public StaticVisitor<bool>
     bool operator() (const Int64 & l, const Tuple & r)      const { return cantCompare(l, r); }
     bool operator() (const Int64 & l, const AggregateFunctionStateData & r) const { return cantCompare(l, r); }
 
-    bool operator() (const Float64 & l, const Null & r)     const { return cantCompare(l, r); }
+    bool operator() (const Float64 &, const Null &)         const { return NULLS_LAST; }
     bool operator() (const Float64 & l, const UInt64 & r)   const { return accurate::lessOp(l, r); }
     bool operator() (const Float64 & l, const UInt128 & r)  const { return cantCompare(l, r); }
     bool operator() (const Float64 & l, const Int64 & r)    const { return accurate::lessOp(l, r); }
```
Comment (alexey-milovidov) [inline dbms/src/Common/FieldVisitors.h]: Also look at line 342, maybe it was incorrect.
```
template <typename T>
bool operator() (const Null &, const T &) const
```

Comment (zlobober) [reply dbms/src/Common/FieldVisitors.h]: Yes, I forgot about this method, thanks.

**Review by zlobober — COMMENTED** [review]
We discussed introduction of weak symbol NULLS_LAST internally with @alexey-milovidov and decided to continue with NULLS_FIRST comparison here without any possible customization.

**Comment by zlobober** [comment]
@alexey-milovidov, would you mind checking again?

File: dbms/src/Common/FieldVisitors.h [hunk 2]
```diff
@@ -372,6 +389,8 @@ class FieldVisitorAccurateLess : public StaticVisitor<bool>
             return l < r;
         else if constexpr (std::is_same_v<U, Int64> || std::is_same_v<U, UInt64>)
             return l < DecimalField<Decimal128>(r, 0);
+        if constexpr (std::is_same_v<T, Null>)
```
Comment (alexey-milovidov) [inline dbms/src/Common/FieldVisitors.h]: It looks wrong.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by alexey-milovidov** [comment]
Request for the next PR:
- add a setting `allow_nullable_key`, false by default;
- if it is set or if we attach table, allow Nullable key;
- add a test where comparison with null is covered.

File: dbms/src/Common/FieldVisitors.h (hunk 2 above)
Comment (zlobober) [reply dbms/src/Common/FieldVisitors.h]: :( You are right, I misinterpreted the meaning of T.

**Comment by zlobober** [comment]
One more attempt.

**Comment by calebeaires** [comment]
Hey @alexey-milovidov, is your solution implemented? If so, who take it on the instance!?

**Comment by amosbird** [comment]
It seems nullable keys are valid after this pr. I'll help finish what @alexey-milovidov proposed

**Comment by alexey-milovidov** [comment]
Ok. And please add a setting to alllow/disallow Nullable in primary key, because allowing them contradicts with SQL standard or usual behaviour of other DBMS.

**Pull request merged**